File: app/src/main/java/com/faendir/lightning_launcher/multitool/proxy/ProxyFactory.java
package com.faendir.lightning_launcher.multitool.proxy;

import com.faendir.lightning_launcher.multitool.util.LightningObjectFactory.EvalFunction;
import java9.util.stream.Stream;

import java.lang.reflect.Array;
import java.lang.reflect.InvocationHandler;
import java.lang.reflect.Method;
import java.util.Arrays;

/**
 * @author lukas
 * @since 04.07.18
 */
public final class ProxyFactory {
    private ProxyFactory() {
    }

    static <T extends Proxy> T lightningProxy(Object lightningObject, Class<T> interfaceClass) {
        //noinspection unchecked
        return (T) java.lang.reflect.Proxy.newProxyInstance(ProxyFactory.class.getClassLoader(), new Class[]{interfaceClass}, new JavaProxyInvocationHandler(lightningObject));
    }

    static Lightning evalProxy(EvalFunction eval) {
        return (Lightning) java.lang.reflect.Proxy.newProxyInstance(ProxyFactory.class.getClassLoader(), new Class[]{Lightning.class}, new EvalProxyInvocationHandler(eval));
    }

    public static <T extends Proxy> T cast(Proxy proxy, Class<T> interfaceClass) {
        return lightningProxy(proxy.getReal(), interfaceClass);
    }

    private static class JavaProxyInvocationHandler extends BaseProxyInvocationHandler {
        private final Class<?> clazz;
        private final Object invokeOn;

        JavaProxyInvocationHandler(Object lightningObject) {
            super(lightningObject);
            boolean isClass = lightningObject instanceof Class;
            clazz = isClass ? (Class<?>) lightningObject : lightningObject.getClass();
            invokeOn = isClass ? null : lightningObject;
        }

        @Override
        protected Object doInvoke(String methodName, Class<?>[] parameterTypes, Object[] parameters) throws Exception {
            return clazz.getMethod(methodName, parameterTypes).invoke(invokeOn, parameters);
        }
    }

    private static class EvalProxyInvocationHandler extends BaseProxyInvocationHandler {
        private final EvalFunction eval;

        private EvalProxyInvocationHandler(EvalFunction eval) {
            super(eval);
            this.eval = eval;
        }

        @Override
        protected Object doInvoke(String methodName, Class<?>[] parameterTypes, Object[] parameters) {
            return eval.eval(null, methodName, parameters);
        }
    }

    private abstract static class BaseProxyInvocationHandler implements InvocationHandler {
        private final Object object;

        protected BaseProxyInvocationHandler(Object object) {
            this.object = object;
        }

        @Override
        public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
            if ("getReal".equals(method.getName())) {
                return object;
            }
            Object result;
            Class<?>[] parameterTypes = method.getParameterTypes();
            try {
                if (args == null) {
                    args = new Object[0];
                }
                for (int i = 0; i < parameterTypes.length; i++) {
                    Class<?> type = parameterTypes[i];
                    if (Proxy.class.isAssignableFrom(type)) {
                        args[i] = ((Proxy) args[i]).getReal();
                        if (Proxy.class.isAssignableFrom(type)) {
                            Class<?> t = args[i].getClass();
                            while (!t.getSimpleName().equals(type.getSimpleName())) {
                                t = t.getSuperclass();
                            }
                            parameterTypes[i] = t;
                        }
                    } else if (type.getName().startsWith("org.mozilla.javascript")) {
                        parameterTypes[i] = object.getClass().getClassLoader().loadClass(type.getName());
                    }
                }
                result = doInvoke(method.getName(), parameterTypes, args);
            } catch (NoSuchMethodException e) {
                throw new NoSuchMethodException(object.getClass().getName() + "#" + method.getName() + " " + Arrays.toString(parameterTypes));
            }
            if (result != null) {
                if (Proxy.class.isAssignableFrom(method.getReturnType())) {
                    //noinspection unchecked
                    result = lightningProxy(result, (Class<? extends Proxy>) method.getReturnType());
                } else if (method.getReturnType().isArray() && Proxy.class.isAssignableFrom(method.getReturnType().getComponentType())) {
                    //noinspection unchecked
                    Class<? extends Proxy> componentType = (Class<? extends Proxy>) method.getReturnType().getComponentType();
                    result = Stream.of((Object[]) result).map(object -> lightningProxy(object, componentType)).toArray(i -> (Object[]) Array.newInstance(componentType, i));
                }
            }
            return result;
        }

        protected abstract Object doInvoke(String methodName, Class<?>[] parameterTypes, Object[] parameters) throws Exception;
    }
}

File: app/src/main/java/com/faendir/lightning_launcher/multitool/proxy/ProxyFactory.kt
package com.faendir.lightning_launcher.multitool.proxy

import com.faendir.lightning_launcher.multitool.util.LightningObjectFactory.EvalFunction
import java.lang.reflect.InvocationHandler
import java.lang.reflect.Method
import java.util.*

/**
 * @author lukas
 * @since 04.07.18
 */
object ProxyFactory {

    internal fun <T : Proxy> lightningProxy(lightningObject: Any, interfaceClass: Class<T>): T {
        @Suppress("UNCHECKED_CAST")
        return java.lang.reflect.Proxy.newProxyInstance(ProxyFactory::class.java.classLoader, arrayOf<Class<*>>(interfaceClass), JavaProxyInvocationHandler(lightningObject)) as T
    }

    internal fun evalProxy(eval: EvalFunction): Lightning {
        return java.lang.reflect.Proxy.newProxyInstance(ProxyFactory::class.java.classLoader, arrayOf<Class<*>>(Lightning::class.java), EvalProxyInvocationHandler(eval)) as Lightning
    }

    fun <T : Proxy> cast(proxy: Proxy, interfaceClass: Class<T>): T {
        return lightningProxy(proxy.real, interfaceClass)
    }

    private class JavaProxyInvocationHandler internal constructor(lightningObject: Any) : BaseProxyInvocationHandler(lightningObject) {
        private val clazz: Class<*>
        private val invokeOn: Any?

        init {
            val isClass = lightningObject is Class<*>
            clazz = if (isClass) lightningObject as Class<*> else lightningObject.javaClass
            invokeOn = if (isClass) null else lightningObject
        }

        @Throws(Exception::class)
        override fun doInvoke(methodName: String, parameterTypes: Array<Class<*>>, parameters: Array<Any?>): Any? = clazz.getMethod(methodName, *parameterTypes).invoke(invokeOn, *parameters)
    }

    private class EvalProxyInvocationHandler internal constructor(private val eval: EvalFunction) : BaseProxyInvocationHandler(eval) {
        override fun doInvoke(methodName: String, parameterTypes: Array<Class<*>>, parameters: Array<Any?>): Any? = eval.eval(methodName, *parameters)
    }

    private abstract class BaseProxyInvocationHandler protected constructor(private val `object`: Any) : InvocationHandler {

        @Throws(Throwable::class)
        override fun invoke(proxy: Any, method: Method, arguments: Array<Any?>?): Any? {
            val args = arguments ?: arrayOfNulls(0)
            if ("getReal" == method.name) {
                return `object`
            }
            var result: Any?
            val parameterTypes = method.parameterTypes
            try {
                for (i in parameterTypes.indices) {
                    val type = parameterTypes[i]
                    if (Proxy::class.java.isAssignableFrom(type)) {
                        args[i] = (args[i] as Proxy).real
                        parameterTypes[i] = findClassWithSimpleNameInHierarchy(type.simpleName, args[i]?.javaClass)
                    }
                }
                result = doInvoke(method.name, parameterTypes, args)
            } catch (e: NoSuchMethodException) {
                throw NoSuchMethodException(`object`.javaClass.name + "#" + method.name + " " + Arrays.toString(parameterTypes))
            }

            if(result != null) {
                if (Proxy::class.java.isAssignableFrom(method.returnType)) {
                    @Suppress("UNCHECKED_CAST")
                    result = lightningProxy(result, method.returnType as Class<out Proxy>)
                } else if (method.returnType.isArray && Proxy::class.java.isAssignableFrom(method.returnType.componentType!!)) {
                    @Suppress("UNCHECKED_CAST")
                    val componentType = method.returnType.componentType as Class<out Proxy>
                    @Suppress("UNCHECKED_CAST")
                    result = (result as Array<Any>).map { `object` -> lightningProxy(`object`, componentType) }.toTypedArray()
                }
            }
            return result
        }

        @Throws(Exception::class)
        protected abstract fun doInvoke(methodName: String, parameterTypes: Array<Class<*>>, parameters: Array<Any?>): Any?

        private fun findClassWithSimpleNameInHierarchy(name: String, check: Class<*>?): Class<*>? {
            when {
                check == null -> return null
                check.simpleName == name -> return check
                else -> {
                    for (i in check.interfaces) {
                        val result = findClassWithSimpleNameInHierarchy(name, i)
                        if (result != null) {
                            return result
                        }
                    }
                    return findClassWithSimpleNameInHierarchy(name, check.superclass)
                }
            }
        }
    }
}
